### ISSUE TITLE
Everytime after uploading something, I get: Unfortunately, MozStumbler has stopped

### ISSUE THREAD

**Issue by derickr** [mozilla/MozStumbler#614]
Since May 13th, (I know it was that date, because that's the last successful upload date), after every time it uploads 50 "Observations", I get the error "MozStumbler has stopped". If I go back into the app, and click on the "upload" icon again, this happens again. The last upload time never updates, and it looks like no automated updates happen either.

**Comment by hannosch** [comment]
Which version of MozStumbler are you using (in the about menu) and where did you install it from? Github release pages or f-droid catalog or somewhere else?

**Comment by derickr** [comment]
0.20.4 — I installed the original one through github, but that was one of the first versions. Since then it has updated through its own updating mechanism.

**Comment by Djfe** [comment]
Probably #598 and #606
Will be fixed in the next version.

The app is crashing because the phone type is not correctly (empty instead of gsm/cdma).
Which phone are you using?

**Comment by derickr** [comment]
It's a samsung galaxy s4

**Comment by Djfe** [comment]
Can you provide a catlog?

**Comment by derickr** [comment]
How do I do that? I know nothing about Android debugging. I'm just a user. Happy to help getting one though.

**Comment by Djfe** [comment]
there are lot's of other tutorials on xda, but this one should work ;) :
http://forum.xda-developers.com/showthread.php?t=1616059

if you have questions, just ask them here ;)

**Comment by derickr** [comment]
All right, I managed that. I also upgrade to 0.20.5, and there is the same problem. The log includes both the errors for 0.20.4 and 0.20.5 (and they look the same).

I've put the log at http://derickrethans.nl/files/dump/logcat.5.txt

It looks like I am getting an outofmemory error:

```
D/dalvikvm( 7782): GC_CONCURRENT freed 1835K, 33% free 18284K/27036K, paused 2ms+4ms, total 50ms
D/dalvikvm( 7546): GC_BEFORE_OOM freed 77K, 17% free 102930K/123620K, paused 597ms, total 597ms
E/dalvikvm-heap( 7546): Out of memory on a 12712986-byte allocation.
I/dalvikvm( 7546): "SyncAdapterThread-1" prio=5 tid=15 RUNNABLE
I/dalvikvm( 7546):   | group="main" sCount=0 dsCount=0 obj=0x42d21750 self=0x7b7c5018
I/dalvikvm( 7546):   | sysTid=7665 nice=10 sched=0/0 cgrp=apps/bg_non_interactive handle=2071745648
I/dalvikvm( 7546):   | state=R schedstat=( 0 0 0 ) utm=3561 stm=228 core=2
I/dalvikvm( 7546):   at java.lang.String.<init>(String.java:~422)
I/dalvikvm( 7546):   at java.lang.AbstractStringBuilder.toString(AbstractStringBuilder.java:642)
I/dalvikvm( 7546):   at java.lang.StringBuilder.toString(StringBuilder.java:663)
I/dalvikvm( 7546):   at org.json.JSONStringer.toString(JSONStringer.java:430)
I/dalvikvm( 7546):   at org.json.JSONObject.toString(JSONObject.java:642)
I/dalvikvm( 7546):   at org.mozilla.mozstumbler.sync.SyncAdapter.getRequestBody(SyncAdapter.java:231)
I/dalvikvm( 7546):   at org.mozilla.mozstumbler.sync.SyncAdapter.onPerformSync(SyncAdapter.java:91)
I/dalvikvm( 7546):   at android.content.AbstractThreadedSyncAdapter$SyncThread.run(AbstractThreadedSyncAdapter.java:259)
I/dalvikvm( 7546):
W/dalvikvm( 7546): threadid=15: thread exiting with uncaught exception (group=0x418c6da0)
E/AndroidRuntime( 7546): FATAL EXCEPTION: SyncAdapterThread-1
E/AndroidRuntime( 7546): Process: org.mozilla.mozstumbler, PID: 7546
E/AndroidRuntime( 7546): java.lang.OutOfMemoryError
E/AndroidRuntime( 7546):    at java.lang.String.<init>(String.java:422)
E/AndroidRuntime( 7546):    at java.lang.AbstractStringBuilder.toString(AbstractStringBuilder.java:642)
E/AndroidRuntime( 7546):    at java.lang.StringBuilder.toString(StringBuilder.java:663)
E/AndroidRuntime( 7546):    at org.json.JSONStringer.toString(JSONStringer.java:430)
E/AndroidRuntime( 7546):    at org.json.JSONObject.toString(JSONObject.java:642)
E/AndroidRuntime( 7546):    at org.mozilla.mozstumbler.sync.SyncAdapter.getRequestBody(SyncAdapter.java:231)
E/AndroidRuntime( 7546):    at org.mozilla.mozstumbler.sync.SyncAdapter.onPerformSync(SyncAdapter.java:91)
E/AndroidRuntime( 7546):    at android.content.AbstractThreadedSyncAdapter$SyncThread.run(AbstractThreadedSyncAdapter.java:259)
W/ActivityManager(  810):   Force finishing activity org.mozilla.mozstumbler/.MainActivity
D/KeyguardViewMediator( 1254): setHidden false
D/KeyguardUpdateMonitor( 1254): sendKeyguardVisibilityChanged(true)
D/KeyguardUpdateMonitor( 1254): handleKeyguardVisibilityChanged(1)
D/CrashAnrDetector(  810): processName: org.mozilla.mozstumbler
D/CrashAnrDetector(  810): broadcastEvent : org.mozilla.mozstumbler data_app_crash
```

**Comment by Djfe** [comment]
since you said above that the last successful upload was on May 13th, I guess that you were using v0.19.0 then, right?

So the issue has been introduced with v0.20.0. (released on May 15th)
(see https://github.com/mozilla/MozStumbler/releases)

Maybe someone here has an idea which change could cause this to happen.

I'm wondering what the issue might be since the Galaxy S4 should usually have enough free RAM.

@derickr how many observations/cells/wifis are in your queue at the moment? Does it help to clear your apps data? (if you are fearing to loose the stumbled data than you don't need to test it, unless your device is rooted than you could make a backup)

Does using a sync adapter really makes sense btw.? I know that it was introduced to save power, but are there any other magnificent advantages?

**Comment by derickr** [comment]
It has in queue: 12500 observations, 60760 cells and 5962954 access points. I'd rather not lose that data as it's been collected in slightly remote locations.

**Comment by Djfe** [comment]
quote from an email from the mailing list:

``` Hi,

Since a few days, my MozStumbler app is not able to upload observations anymore. It looks like the application created wrong data: The number of collected cell and wifi observations was ways too high compared to the days and months before. I can still collect new data, but each time the app starts to upload it crashes.

Is there any possibility to analyze the data? Where is the MozStumbler data stored on the phone? Or is the only way to solve this issue to delete the MozStumber data in the Settings/Application Manager?

Regards, Christian
```

I think I know what the issue is now:
it's #591 and got fixed with #592 in v0.20.3 (v0.20.1 and 2 were failed releases and don't count)
but the data that was stumbled in v0.20.0 is still remaining

of course you could just delete the data and everything will work fine
but of course that's not what you want

the reason for your issue is, that the data has been duplicated and increased exponentially (-> the longer the drive the bigger the amount of data that has been added), the data from previous scans remained in the variables that will be saved in the database, so that for each scan all previous scans will be added to the database again

so the following data contains huge amounts of duplications
12500 observations, 60760 cells and 5962954 access points

maybe it is possible to write a workaround just for the next update:
the code will be executed only directly after the update (I think there is already code to update the database, maybe that can be used)
we will just pretend to update the database, but what will happen is that all duplicates will be removed from the database
we can leave the code in if somebody decides to update at a later date (a few versions later), but everybody else won't get it again because they have the "newer" database version

after that there is a significantly smaller amount of data that needs to be concatenated to a json string (no out-of-memory error again) and the upload should work

to sum it up
**somebody needs to write an update script for the internal database that removes all duplicates that where created due to the bug #591 that got fixed in #592**

**Comment by gpoo** [comment]
@Djfe How can someone know which one is duplicated?

If I am in position `a`, and MozStumbler see `X` SSID, when I move to position `b`, MozStumbler still could see `X` SSID, and that would be two valid records.  What would the threshold be in order to consider a data as noise? (maybe if the value for `signal` is the same?)

**Comment by garvankeeley** [comment]
@Djfe in the bug the data was an _exact_ duplicate, many times over.

**Comment by Djfe** [comment]
So do you think we can fix it this way or do we need something that splits up JSON strings in a different way? (a way that considers memory consumption)

Btw. Christian Wulle send me a private email in which he said that Mozstumbler takes up 93mbs of data :O.
That sounds like duplication for me.

**Comment by Djfe** [comment]
@derickr I think this issue has been fixed in 0.21.0 (kinda)
data that is over 4 weeks old gets deleted now, so once you start MozStumbler 0.21.0+ it should delete the faulty data and you can upload again.

If that is not what you want: this is a warning that 0.21.0 will delete this data.

**Comment by derickr** [comment]
It's OK - I've already deleted the data myself, and started collecting and uploading again.

**Comment by Djfe** [comment]
you could close this issue then ;)

**Comment by derickr** [comment]
Right :-)
